viewer/src/main/java/nl/b3p/viewer/stripes/ApplicationActionBean.java

```diff
@@ -411,7 +414,30 @@ public static Resolution checkRestriction(ActionBeanContext context, Application
                 u = em.find(User.class, p.getName());
             }
         }
-        if (!Authorizations.isApplicationReadAuthorized(application, context.getRequest(), em) && (username == null || u != null && u.isAuthenticatedByIp())) {
+
+        // Check if user has expire date
+        Date today = null;
+        Date expire = null;
+        try {
+            DateFormat formatter = new SimpleDateFormat("dd-MM-yyyy");
+            expire = formatter.parse(u.getDetails().getOrDefault("expiry_date",formatter.format(new Date())));
+            today = formatter.parse(formatter.format(new Date()));
+        } catch (ParseException e){
+            ResourceBundle bundle = ResourceBundleProvider.getResourceBundle(determineLocaleForBundle(context, application));
+            String msg = bundle.getString("viewer.applicationactionbean.expired_parse_error");
+            context.getValidationErrors().addGlobalError(new SimpleError(msg));
+            context.getRequest().getSession().invalidate();
+            return new ForwardResolution("/WEB-INF/jsp/error.jsp");
+        }
+
+        if(today.after(expire)){
+            ResourceBundle bundle = ResourceBundleProvider.getResourceBundle(determineLocaleForBundle(context, application));
+            String msg = bundle.getString("viewer.applicationactionbean.expired");
+            context.getValidationErrors().addGlobalError(new SimpleError(msg));
+            context.getRequest().getSession().invalidate();
+            return new ForwardResolution("/WEB-INF/jsp/error.jsp");
+        }
+        else if (!Authorizations.isApplicationReadAuthorized(application, context.getRequest(), em) && (username == null || u != null && u.isAuthenticatedByIp())) {
```

Comment:
I think it would be better to place this in [Authorizations](https://github.com/flamingo-geocms/flamingo/blob/master/viewer-config-persistence/src/main/java/nl/b3p/viewer/config/security/Authorizations.java)

possibly as part of `isReadAuthorized`, maybe a a new method

the parseError should not go to the user.
the login failure should be a generic login failure error and redirect to the loginfail jsp/page, not the error page

Reply:
Yes its better to move the code to its own method in Authorizations.
Not sure if redirecting to the loginfail page is better, because the user gets a default message that the login information is incorrect what is not the case here. Besides the user can try to login again, this is useless because the account has expired.